framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/log/InstancesLogCollectorRunnable.java

```diff
@@ -0,0 +1,103 @@
+package org.kie.cloud.openshift.log;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.kie.cloud.api.deployment.Instance;
+import org.kie.cloud.common.logs.InstanceLogUtil;
+import org.kie.cloud.openshift.deployment.OpenShiftInstance;
+import org.kie.cloud.openshift.resource.Project;
+import org.kie.cloud.openshift.scenario.OpenShiftScenario;
+import org.kie.cloud.openshift.util.OpenshiftInstanceUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InstancesLogCollectorRunnable implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftScenario.class);
+
+    private static final Integer DEFAULT_OBERVABLE_BUFFER_IN_SECONDS = 5;
+
+    private Project project;
+    private String logFolderName;
+
+    private ExecutorService executorService = Executors.newCachedThreadPool();
+    private Set<OpenShiftInstance> observedInstances = Collections.synchronizedSet(new HashSet<>());
+
+    public InstancesLogCollectorRunnable(Project project, String logFolderName) {
+        super();
+        this.project = project;
+        this.logFolderName = logFolderName;
+    }
+
+    @Override
+    public void run() {
+        // Check for new instances and observe on them
+        OpenshiftInstanceUtil.getAllInstances(project.getOpenShift(), project.getName()).stream()
+                             // Filter non observed instances
+                             .filter(instance -> !isInstanceObserved(instance))
+                             // Observe instance logs
+                             .forEach(this::observeInstanceLog);
+
+        // Check if instances are still existing
+        // Calling get pod on a no more existing pod will cause the "obervePodLog" to complete ...
+    }
+
+    public void closeAndFlushRemainingInstanceCollectors() {
+        // Keep instances list before stopping everything
+        List<OpenShiftInstance> instances = new ArrayList<>(observedInstances);
+        // Stop all observations
+        executorService.shutdownNow();
+        // Flush all remaining instances (which were still observed)
+        instances.forEach(this::flushInstanceLogs);
```

Comment:
If you flush instance logs here, doesn't it duplicate already existing log content of instances which are being observed?

Reply:
The log content will be overridden .

Reply:
True. It is to make sure we have the latest state of running pods.